### PR TITLE
Add Cuneiform programming language

### PR DESCRIPTION
**Pull request by joergen7**
This PR adds an entry for Cuneiform in `languages.yml`. Also, it adds a few code samples to the sample directory.
- [cuneiform-lang.org](https://cuneiform-lang.org/)
- [joergen7/cuneiform](https://github.com/joergen7/cuneiform)
- [Cuneiform (programming language)](https://en.wikipedia.org/wiki/Cuneiform_(programming_language))

## Checklist:

- [x] **I am associating a language with a new file extension.**
  - [ ] The new extension is used in hundreds of repositories on GitHub.com
    - Search results for each extension:
      
      - https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3Acfl+def+Str+Bool+File+NOT+nothack
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - [ackermann.cfl](samples/Cuneiform/ackermann.cfl)
      - [r_test.cfl](samples/Cuneiform/r_test.cfl)
      - [fizzbuzz_python.cfl](samples/Cuneiform/fizzbuzz_python.cfl)
    - Sample license(s):
  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.

- [ ] **I am adding a new language.**
  - [ ] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
      
      -  https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3AFOOBAR+KEYWORDS+NOT+nothack
  - [ ] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - [URL to each sample source, if applicable]
    - Sample license(s):
  - [ ] I have included a syntax highlighting grammar.
  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.

- [ ] **I am fixing a misclassified language**
  - [ ] I have included a new sample for the misclassified language:
    - Sample source(s):
      - [URL to each sample source, if applicable]
    - Sample license(s):
  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.

- [ ] **I am changing the source of a syntax highlighting grammar**
  
  - Old: https://github-lightshow.herokuapp.com/
  - New: https://github-lightshow.herokuapp.com/

- [ ] **I am adding new or changing current functionality**
  
  - [ ] I have added or updated the tests for the new or changed functionality.

### CODE REVIEW

**Comment by Alhadis** [comment]
I'm thoroughly disappointed this isn't written in 𒀝𒅗𒁺𒌑, or even in Ugaritic. 😢

**Comment by joergen7** [comment]
I admit I have difficulties understanding why some of the tests fail. Most of the failures seem to all originate in `test_language.rb` which I did not touch at all.

**Comment by Alhadis** [comment]
Alllrighty, putting aside my language-obsessed nerdiness, let's have a look at these build results:

~~~
  1) Failure:
TestGrammars#test_submodules_have_approved_licenses [/home/travis/build/github/linguist/test/test_grammars.rb:115]:
The following submodules have unapproved licenses:
* vendor/grammars/FreeMarker.tmbundle: b81acf2ba52d312754bf5055845a723123bda388
* vendor/grammars/ant.tmbundle: 4da01d631a29c76456fd0bd16749c71e8d5f6dbf
* vendor/grammars/elixir-tmbundle: 220e011c8d686129e9c4163a7c655b9d64f61e59
* vendor/grammars/mako-tmbundle: 39f092c726491ca6a02354dbc6c3a0920bb44d4c
The license must be added to the LICENSE_WHITELIST in /test/test_grammars.rb once approved.
--- expected
+++ actual
@@ -1 +1 @@
-[]
+["vendor/grammars/FreeMarker.tmbundle: b81acf2ba52d312754bf5055845a723123bda388", "vendor/grammars/ant.tmbundle: 4da01d631a29c76456fd0bd16749c71e8d5f6dbf", "vendor/grammars/elixir-tmbundle: 220e011c8d686129e9c4163a7c655b9d64f61e59", "vendor/grammars/mako-tmbundle: 39f092c726491ca6a02354dbc6c3a0920bb44d4c"]
  2) Failure:
TestGrammars#test_whitelisted_hashes_dont_have_licenses [/home/travis/build/github/linguist/test/test_grammars.rb:130]:
The following whitelisted license hashes are unused:
* 7dfce11e2e3579ee43b83e69b1b64e77a2e378f0
* 62b97e52b78439c14550a44a3fe51332aeffb3a1
* 0acff2bb1536a3942a39ac74987ffd9c44905a6b
* 41cdc7e9f9d2e62eb8ac68a1a9359b9c39a7a9bf
Please remove them from the hash whitelist.
--- expected
+++ actual
@@ -1 +1 @@
~~~

These are unrelated to your real changes, and suggest that your fork is a few releases behind upstream. Specifically, your submodules need updating. Try running `script/bootstrap` and manually deleting any unlisted modules with `rm -rf vendor/grammars/etc`.

If that doesn't work, it might help to trash everything and start from a fresh fork. 👍

File: lib/linguist/languages.yml [hunk 1]
```diff
@@ -946,6 +936,17 @@ Cuda:
...
+  type: programming
+  ace_mode: text
+  extensions:
+  - ".cfl"
+  interpreters:
+  - "cuneiform"
+  - "cf_client"
+  color: "#501616"
+  tm_scope: none
+  language_id: 2699
```
[the remark this answers is not in the document]
Comment (joergen7) [reply lib/linguist/languages.yml]: No, I picked a number that did not appear in the yml file :D
will correct ...

File: lib/linguist/languages.yml [hunk 2]
```diff
@@ -615,6 +615,18 @@ CSV:
...
+  type: programming
+  extensions:
+  - ".cf"
+  - ".cfl"
+  - ".cuneiform"
+  interpreters:
+  - "cuneiform"
+  - "cf_client"
+  color: "#2772b0"
+  tm_scope: none
```
Comment (Alhadis) [inline lib/linguist/languages.yml]: Even though it's a different language, you can use Ruby's grammar for Cuneiform highlighting. The results [look decent enough](https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=source.ruby&grammar_format=auto&grammar_url=&grammar_text=&code_source=from-text&code_url=&code=def+inc%28+x+%3A+Str+%29+-%3E+Str+%7B%0D%0A%0D%0A++def+inc%28+x+%3A+Str+%29+-%3E+%3Cy+%3A+Str%3E+in+Python+*%7B%0D%0A++++y+%3D+int%28+x+%29%2B1%0D%0A++%7D*%0D%0A%0D%0A++%28+inc%28+x+%3D+x+%29%7Cy+%29%0D%0A%7D%0D%0A%0D%0Adef+dec%28+x+%3A+Str+%29+-%3E+Str+%7B%0D%0A%0D%0A++def+dec%28+x+%3A+Str+%29+-%3E+%3Cy+%3A+Str%3E+in+Python+*%7B%0D%0A++++y+%3D+int%28+x+%29-1%0D%0A++%7D*%0D%0A%0D%0A++%28+dec%28+x+%3D+x+%29%7Cy+%29%0D%0A%7D).

If that isn't good enough, just give me the formal language specification, and I'll write a grammar for you. 😉

**Comment by joergen7** [comment]
Even if my installation is broken, I don't understand how this affects Travis CI. I changed one file and added 4 samples, nothing more.

**Comment by joergen7** [comment]
I have just run

```
bundle exec rake
```

on a clean checkout of the `master` branch and got the same errors.

**Comment by Alhadis** [comment]
> I did this:
> ~~~
> jorgen@x240:~/git/linguist$ git merge upstream/master
> Already up to date.
> ~~~

@joergen7 What does `script/bootstrap` output when running it? If your submodules are in sync, you shouldn't see anything. OTOH, if you see output along the lines of *"Submodule vendor/grammars/… registered for …etc"*, it indicates the locally checked-out submodules haven't been updated.

Submodules are managed independently of the "parent" module, which means it isn't enough to run `git merge` by itself. Sometimes `git submodule init` or `git submodule update` is needed to ensure each module is checked out to the commit registered in `.gitmodules`, and so forth. :)

**Comment by joergen7** [comment]
Running
```
script/bootstrap
```
outputs nothing the second time I run it. (The first time it is quite verbose, should I grep?)

**Comment by jonabc** [comment]
I don't believe that the latest version of `licensed` has been pulled into `linguist` yet, so https://github.com/github/linguist/pull/4152#issuecomment-394423156 shouldn't be an issue.  You might be hitting similar issues to what was found in https://github.com/github/linguist/pull/4139 though?

**Comment by Alhadis** [comment]
> The first time it is quite verbose, should I grep?

What do you see when you run `git status`? And are you still getting test failures?

**Comment by joergen7** [comment]
Running `git status` on my current pull request:
```
jorgen@x240:~/git/linguist$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```

Also when I build from a clean clone of linguist I get the following output:
[Output of running scripts bootstrap on a clean checkout of linguist ](https://gist.github.com/joergen7/0373d2aa038e0825014ab4f22cf5a72d)

**Comment by Alhadis** [comment]
I'm afraid I'm tapped. :( I'll leave this up to @lildude... @pchaigno might also know what's going on here..

**Comment by joergen7** [comment]
I guess I'll just wait for a fix.

In the mean time: Do you think I can use the [grammar-spec](https://github.com/joergen7/cuneiform-syntax-highlight/blob/master/Cuneiform.sublime-syntax) I created for Sublime-text here? Is there a hint which grammar specs are ok, naming conventions for directories, etc?

**Comment by Alhadis** [comment]
Erm, I'm not sure. I recall that we [had troubles](https://github.com/github/linguist/pull/3214) a while ago with parsing `.sublime-syntax` grammars, and I don't believe the situation's changed. Sublime's grammar format is incompatible with the older TextMate-based format, so I'm leaning on an answer of "no". 😉

**Comment by joergen7** [comment]
That's fine. Now I know I have to build an ace-grammar. But that's good for ace users, I guess.

**Comment by Alhadis** [comment]
Although your grammar is pretty self-explanatory... I'm not seeing any of Sublime's "incompatible" grammar features (the pushing and popping thing). Give me a moment, I'll convert it to a CSON file which you can stick in a `grammars/` subdirectory (which is the conventional location for Atom grammars).

**Comment by joergen7** [comment]
It pushes for `string` and pops for `inside_string`. Same for `file`/`inside_file`.

**Comment by Alhadis** [comment]
Done. See joergen7/cuneiform-syntax-highlight#1 =)

Now, adding a grammar to Linguist is simply a case of running the [`script/add-grammar`](https://github.com/github/linguist/blob/master/CONTRIBUTING.md#adding-a-language) script. You'd run it like this:

~~~console
$ script/add-grammar joergen7/cuneiform-syntax-highlight

# Or like this, both notations are supported.
$ script/add-grammar https://github.com/joergen7/cuneiform-syntax-highlight
~~~

**Comment by pchaigno** [comment]
The Travis CI build failure is unrelated to this pull request and we'll have to fix it on master.

In any case, using [this search query](https://github.com/search?p=1&q=extension%3Acfl+def+OR+Str+OR+Bool&type=Code&utf8=%E2%9C%93), I count only 6 users of that file extension for Cuneiform. That won't be enough to meet your hundreds-of-repositories criteria.

**Comment by joergen7** [comment]
@pchaigno Thanks for having considered my pull request, even though it did not meet the user-threshold criterion.

I will just close this pull request and re-open it, should we manage to meet all necessary criteria.

**Comment by pchaigno** [comment]
> I will just close this pull request and re-open it, should we manage to meet all necessary criteria.

:+1:

**Pull request closed without merging**